plugins/koterm.koplugin/main.lua

```diff
@@ -0,0 +1,575 @@
+        self.input_widget:trimBuffer(self.buffer_size)
+        if self.is_shell_open then
+            UIManager:tickAfterNext(function()
+                UIManager:scheduleIn(self.refresh_time, Terminal.refresh, self)
+            end)
+        end
+    else
+        if self.is_shell_open then
+            if self.refresh_time > 5 then
+                self.refresh_time = self.refresh_time
+            elseif self.refresh_time > 1 then
+                self.refresh_time = self.refresh_time * 1.1
+            else
+                self.refresh_time = self.refresh_time * 2
+            end
+            UIManager:scheduleIn(self.refresh_time, Terminal.refresh, self)
+        end
+    end
+end
+
+function Terminal:transmit(chars)
+    C.write(self.ptmx, chars, #chars)
+    self:refresh(true)
+end
+
+--- kills a running shell
+-- @param ask if true ask if a shell is running, don't kill
+-- @return pid if shell is running, -1 otherwise
+function Terminal:killShell(ask)
+    UIManager:unschedule(Terminal.refresh)
+    local pid_file = io.open("koterm.pid", "r")
+    if not pid_file then
+        return -1
+    end
+
+    local pid = tonumber(pid_file:read("*a"))
+    pid_file:close()
+
+    if ask then
+        return pid
+    else
+        local terminate = "\03\n\nexit\n"
+        self:transmit(terminate)
+        -- do other things before killing first
+        self.is_shell_open = false
+        self.history = ""
+        os.remove("koterm.pid")
+        C.close(self.ptmx)
+
+        C.kill(pid, C.SIGTERM)
+
+        local status = ffi.new('int[1]')
+        -- status = tonumber(status[0])
+        -- If still running: ret = 0 , status = 0
+        -- If exited: ret = pid , status = 0 or 9 if killed
+        -- If no more running: ret = -1 , status = 0
+        C.waitpid(pid, status, 0) -- wait until shell is terminated
+
+        return -1
+    end
+end
+
+function Terminal:getCharSize()
+    local tmp = TextWidget:new{
+        text = " ",
+        face = self.input_face,
+    }
+    return tmp:getSize().w
+end
+
+function Terminal:generateInputDialog()
+    return InputDialog:new{
+        title =  _("Terminal Emulator"),
+        input = self.history,
+        input_face = self.input_face,
+        para_direction_rtl = false,
+        input_type = "string",
+        allow_newline = false,
+        cursor_at_end = true,
+        fullscreen = true,
+        inputtext_class = TermInputText,
+        buttons = {{
+            {
+            text = "↹",  -- tabulator "⇤" and "⇥"
+            callback = function()
+                self:transmit("\009")
+            end,
+            },
+            {
+            text = "/",  -- slash
+            callback = function()
+                self:transmit("/")
+            end,
+            },
+            {
+            text = _("Esc"),
+            callback = function()
+                self:transmit("\027")
+            end,
+            },
+            {
+            text = _("Ctrl"),
```

Comment:
Do these keys really get translated in non-english keyboards ?

Reply:
Yes, in German that's <kbd>Strg</kbd> (Steuerung).

Reply:
It might need a context though, like "keyboard" or something.